src/transformers/models/pop2piano/tokenization_pop2piano.py

```diff
@@ -0,0 +1,428 @@
+        elif token >= self.vocab_size_special:
+            type = self.encoder["TOKEN_NOTE"]
+            value = int(token - self.vocab_size_special)
+        else:
+            type = self.encoder["TOKEN_SPECIAL"]
+            value = int(token)
+
+        return [type, value]
+
+    def _convert_token_to_id(self, token, token_type="3"):
+        if token_type == self.encoder["TOKEN_TIME"]:
+            return self.vocab_size_special + self.vocab_size_note + self.vocab_size_velocity + token
+        elif token_type == self.encoder["TOKEN_VELOCITY"]:
+            return self.vocab_size_special + self.vocab_size_note + token
+        elif token_type == self.encoder["TOKEN_NOTE"]:
+            return self.vocab_size_special + token
+        elif token_type == self.encoder["TOKEN_SPECIAL"]:
+            return token
+        else:
+            return -1
+
+    def relative_batch_tokens_to_midi(
+        self,
+        tokens,
+        beatstep,
+        beat_offset_idx=None,
+        bars_per_batch=None,
+        cutoff_time_idx=None,
+    ):
+        """Converts tokens to midi"""
+
+        beat_offset_idx = 0 if beat_offset_idx is None else beat_offset_idx
+        notes = None
+        bars_per_batch = 2 if bars_per_batch is None else bars_per_batch
+
+        N = len(tokens)
+        for index in range(N):
+            _tokens = tokens[index]
+            _start_idx = beat_offset_idx + index * bars_per_batch * 4
+            _cutoff_time_idx = cutoff_time_idx + _start_idx
+            _notes = self.relative_tokens_to_notes(
+                _tokens,
+                start_idx=_start_idx,
+                cutoff_time_idx=_cutoff_time_idx,
+            )
+
+            if len(_notes) == 0:
+                pass
+            elif notes is None:
+                notes = _notes
+            else:
+                notes = np.concatenate((notes, _notes), axis=0)
+
+        if notes is None:
+            notes = []
+        midi = self.notes_to_midi(notes, beatstep, offset_sec=beatstep[beat_offset_idx])
+        return midi, notes
+
+    # Taken from the original code
+    # Please see https://github.com/sweetcocoa/pop2piano/blob/fac11e8dcfc73487513f4588e8d0c22a22f2fdc5/midi_tokenizer.py#L257
+    def relative_tokens_to_notes(self, tokens, start_idx, cutoff_time_idx=None):
+        if tokens[0] >= (
+            self.vocab_size_special + self.vocab_size_note + self.vocab_size_velocity + self.vocab_size_time
+        ):
+            tokens = tokens[1:]
+
+        words = [self._convert_id_to_token(token, time_idx_offset=0) for token in tokens]
+
+        if hasattr(start_idx, "item"):
+            """if numpy or torch tensor"""
+            start_idx = start_idx.item()
+
+        current_idx = start_idx
+        current_velocity = 0
+        note_onsets_ready = [None for i in range(self.vocab_size_note + 1)]
+        notes = []
+        for type, number in words:
+            if type == self.encoder["TOKEN_SPECIAL"]:
```

Comment:
The logic in the function is quite complex. It would benefit from splitting up into separate (private) functions that handle the logic for each token type